pkg/server/server_controller.go

```diff
@@ -172,20 +175,146 @@ func newServerController(
+// that should be started during server initialization.
+func (c *serverController) startInitialSecondaryTenantServers(
+	ctx context.Context, ie isql.Executor,
+) error {
+	// The list of tenants that should have a running server.
+	reqTenants, err := c.getExpectedRunningTenants(ctx, ie)
+	if err != nil {
+		return err
+	}
+	for name := range reqTenants {
+		if name == catconstants.SystemTenantName {
+			// We already pre-initialize the entry for the system tenant.
+			continue
+		}
+		if _, err := c.createServer(ctx, name, base.TestSharedProcessTenantArgs{}); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// updateRunningTenants checks which tenants need to be started/stopped
+// and performs the necessary server lifecycle changes.
+func (c *serverController) updateRunningTenants(ctx context.Context, ie isql.Executor) error {
+	// The list of tenants that should have a running server.
+	reqTenants, err := c.getExpectedRunningTenants(ctx, ie)
+	if err != nil {
+		return err
+	}
+
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	// First check if there are any servers that shouldn't be running right now.
+	for name, srv := range c.mu.servers {
+		if !srv.shouldStop {
+			continue
+		}
+		if _, ok := reqTenants[name]; !ok {
+			// Tenant currently running, but should not be running. Shut it down.
+			if err := c.stopper.RunAsyncTask(ctx, "stop-tenant-async", func(ctx context.Context) {
+				srv.server.stop(context.Background())
+			}); err != nil {
+				// Failed to add task: shutting down.
+				return err
+			}
+		}
+	}
+
+	// Now add all the missing servers.
+	for name := range reqTenants {
+		if _, ok := c.mu.servers[name]; !ok {
+			// Server not currently running, and should be running. Start it.
+			tenantName := name // copy loop iteration variable.
+			if err := c.stopper.RunAsyncTask(ctx, "start-tenant-async", func(ctx context.Context) {
+				if _, err := c.createServer(ctx, tenantName, base.TestSharedProcessTenantArgs{}); err != nil {
+					log.Warningf(ctx, "unable to start service for tenant: %v", err)
+				}
+			}); err != nil {
+				// Failed to add task: shutting down.
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// getExpectedRunningTenants retrieves the tenant IDs that should
+// be running right now.
+// TODO(knz): Use a watcher here.
+// Probably as followup to https://github.com/cockroachdb/cockroach/pull/95657.
+func (c *serverController) getExpectedRunningTenants(
+	ctx context.Context, ie isql.Executor,
+) (map[roachpb.TenantName]struct{}, error) {
+	if !c.st.Version.IsActive(ctx, clusterversion.V23_1TenantNamesStateAndServiceMode) {
+		// Cluster not yet upgraded - we know there is no secondary tenant
+		// with a name yet.
+		return make(map[roachpb.TenantName]struct{}), nil
+	}
+
+	rows, err := ie.QueryBuffered(ctx, "fetch-running-tenants",
+		nil, /* txn */
+		`SELECT name FROM system.tenants WHERE service_mode = $1 AND name IS NOT NULL`, mtinfopb.ServiceModeShared)
```

Comment:
I know we prevent it on write, but do we want an extra condition here that the data_mode is also in the correct state?